File: modules/native/src/iosurface.cpp
//
// Created by Nikita Shtengauer on 20.11.2023.
//

#include <jni.h>
#include <CoreFoundation/CoreFoundation.h>
#include <CoreGraphics/CGDisplayConfiguration.h>
#include <IOSurface/IOSurface.h>
#include <OpenGL/OpenGL.h>
#include "openglfx.h"


iosfun(jdouble, nGetDisplayDPI)(JNIEnv *, jobject, jdouble x, jdouble y) {
    CGDirectDisplayID display;
    CGGetDisplaysWithPoint(CGPoint{x, y}, 1, &display, nullptr);

    CFStringRef keys[1]{
        kCGDisplayShowDuplicateLowResolutionModes
    };
    CFBooleanRef values[1]{
        kCFBooleanTrue
    };
    CFDictionaryRef dictionary = CFDictionaryCreate(
                nullptr, (const void **) keys, (const void **) values, 1,
                &kCFCopyStringDictionaryKeyCallBacks,
                &kCFTypeDictionaryValueCallBacks);

    CFArrayRef modes = CGDisplayCopyAllDisplayModes(display, dictionary);
    CFRelease(dictionary);

    int actualWidth = 0;
    for(int i = 0; i < CFArrayGetCount(modes); i++){
        CGDisplayModeRef mode = (CGDisplayModeRef)CFArrayGetValueAtIndex(modes, i);

        int width = CGDisplayModeGetPixelWidth(mode);
        if(width == CGDisplayModeGetWidth(mode) && actualWidth < width)
            actualWidth = width;
    }
    return (double)actualWidth / CGDisplayPixelsWide(display);
}

iosfun(jlong, nCreateIOSurface)(JNIEnv *, jobject, jint width, jint height) {
    int bytes = 4;
    CFStringRef keys[3]{
            kIOSurfaceWidth, kIOSurfaceHeight, kIOSurfaceBytesPerElement
    };
    CFNumberRef values[3]{
            CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &width),
            CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &height),
            CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &bytes)
    };

    CFDictionaryRef dictionary = CFDictionaryCreate(
            nullptr,
            (const void **) keys, (const void **) values, 3,
            &kCFCopyStringDictionaryKeyCallBacks,
            &kCFTypeDictionaryValueCallBacks);
    IOSurfaceRef ioSurfaceRef = IOSurfaceCreate(dictionary);
    CFRelease(dictionary);
    return (jlong) ioSurfaceRef;
}

iosfun(void, nReleaseIOSurface)(JNIEnv *, jobject, jlong ioSurfaceRef) {
    CFRelease((IOSurfaceRef) ioSurfaceRef);
}

iosfun(jint, nCGLTexImageIOSurface2D)(JNIEnv *, jobject,
                                      jlong ctx, jint target, jint internalFormat, jint width, jint height, jint format,
                                      jint type,
                                      jlong ioSurfaceRef, jint plane) {
    return (jint) CGLTexImageIOSurface2D(
            (CGLContextObj) ctx,
            (GLenum) target, (GLenum) internalFormat,
            (GLsizei) width, (GLsizei) height,
            (GLenum) format, (GLenum) type,
            (IOSurfaceRef) ioSurfaceRef, (GLuint) plane);
}

iosfun(void, nIOSurfaceLock)(JNIEnv *, jobject, jlong ioSurfaceRef) {
    IOSurfaceLock((IOSurfaceRef) ioSurfaceRef, kIOSurfaceLockReadOnly, nullptr);
}

iosfun(void, nIOSurfaceUnlock)(JNIEnv *, jobject, jlong ioSurfaceRef) {
    IOSurfaceUnlock((IOSurfaceRef) ioSurfaceRef, kIOSurfaceLockReadOnly, nullptr);
}

File: modules/core/src/main/kotlin/com/huskerdev/openglfx/internal/iosurface/IOSurface.kt
package com.huskerdev.openglfx.internal.iosurface

import com.huskerdev.ojgl.GLContext

internal class IOSurface(val width: Int, val height: Int) {

    companion object {
        @JvmStatic external fun nGetDisplayDPI(x: Double , y: Double): Double

        @JvmStatic private external fun nCreateIOSurface(width: Int, height: Int): Long
        @JvmStatic private external fun nReleaseIOSurface(ioSurfaceRef: Long)
        @JvmStatic private external fun nCGLTexImageIOSurface2D(
            ctx: Long, target: Int, internalFormat: Int,
            width: Int, height: Int, format: Int, type: Int,
            ioSurfaceRef: Long, plane: Int): Int
        @JvmStatic private external fun nIOSurfaceLock(ioSurfaceRef: Long)
        @JvmStatic private external fun nIOSurfaceUnlock(ioSurfaceRef: Long)
    }

    private val handle = nCreateIOSurface(width, height)

    fun cglTexImageIOSurface2D(ctx: GLContext, target: Int, internalFormat: Int,
                               format: Int, type: Int,
                               plane: Int): Int =
        nCGLTexImageIOSurface2D(ctx.handle, target, internalFormat, width, height, format, type, handle, plane)

    fun lock() = nIOSurfaceLock(handle)

    fun unlock() = nIOSurfaceUnlock(handle)

    fun dispose() = nReleaseIOSurface(handle)
}